src/Rules.hs

```diff
@@ -20,33 +21,42 @@ import qualified Rules.Program
 import qualified Rules.Register
 import Settings
 import Target
-import UserSettings
 import Utilities
 
 allStages :: [Stage]
-allStages = [minBound ..]
+allStages = [minBound .. maxBound]
 
 -- | This rule calls 'need' on all top-level build targets, respecting the
 -- 'Stage1Only' flag.
 topLevelTargets :: Rules ()
-topLevelTargets = action $ do
-    let libraryPackages = filter isLibrary (knownPackages \\ [libffi])
-    need =<< if stage1Only
-             then do
-                 libs <- concatForM [Stage0, Stage1] $ \stage ->
-                     concatForM libraryPackages $ packageTargets False stage
-                 prgs <- concatForM programsStage1Only $ packageTargets False Stage0
-                 return $ libs ++ prgs ++ inplaceLibCopyTargets
-             else do
-                 targets <- concatForM allStages $ \stage ->
-                     concatForM (knownPackages \\ [libffi]) $
-                        packageTargets False stage
-                 return $ targets ++ inplaceLibCopyTargets
+topLevelTargets = do
+    phony "stage2" $ do
```

Comment:
The top-level module `Rules` defines the default rule. All other rules are defined in `Rules/...` modules, so let's move the new `stage2` rule in a separate module. 

Or is it actually supposed to be the default build rule? Then drop the `stage2` phony, let it be the default.

Reply:
Yeah I think this would be a sane default rule. This is however one of the points I mentionned in my email as one of the things I want to discuss, that are not definitive and on which I would like your opinion. In particular, I haven't examined very closely the changes to the top-level `Rules` module, so I'm all ears :)

Reply:
My understanding is that the only change is that we now put all built artefacts in nicely structured `stageN/bin` and `stageN/lib` directories. If that's all that has changed, then let's just call this build rule default.

Reply:
My question might be silly but... how do we "set" the default rule?

Reply:
https://github.com/snowleopard/hadrian/blob/master/src/Rules.hs#L31-L44 Just `action $ do` instead of using `phony` like here?

Reply:
@alpmestan Yes, it's just `action $ do` -- have a look at how it is implemented in master.

Reply:
OK, I made that rule the default with `action $ do` and it indeed works, hadrian builds a complete stage2 compiler now when no target is specified.